text/0000-expand-math.md

```diff
@@ -0,0 +1,230 @@
+<<trait>> Integer
+class SignedInteger
+<<trait>> SignedInteger
+class UnsignedInteger
+<<trait>> UnsignedInteger
+
+Any <-- Real
+Real <-- FloatingPoint
+Real <-- Integer
+Integer <-- SignedInteger
+Integer <-- UnsignedInteger
+
+FloatingPoint <-- F32
+FloatingPoint <-- F64
+
+UnsignedInteger <-- U8
+UnsignedInteger <-- U16
+UnsignedInteger <-- U32
+UnsignedInteger <-- U64
+UnsignedInteger <-- U128
+UnsignedInteger <-- USize
+UnsignedInteger <-- ULong
+
+SignedInteger <-- I8
+SignedInteger <-- I16
+SignedInteger <-- I32
+SignedInteger <-- I64
+SignedInteger <-- I128
+SignedInteger <-- ISize
+SignedInteger <-- ILong
+```
+
+This RFC introduces four more numeric types: `Rational`, `BigInt`, `BigFloat`, and `Complex`. These fit into the numeric type hierarchy in the following manner:
+
+```mermaid
+classDiagram
+class Any
+<<interface>> Any
+class Real
+<<trait>> Real
+class FloatingPoint
+<<trait>> FloatingPoint
+class Integer
+<<trait>> Integer
+class SignedInteger
+<<trait>> SignedInteger
+class UnsignedInteger
+<<trait>> UnsignedInteger
+
+Any <-- Real
+Any <-- Complex~Real~
+Real <-- FloatingPoint
+Real <-- Integer
+Real <-- Rational~Integer~
+Integer <-- SignedInteger
+Integer <-- UnsignedInteger
+Integer <-- BigInt
+
+FloatingPoint <-- F32
+FloatingPoint <-- F64
+FloatingPoint <-- BigFloat
+
+UnsignedInteger <-- U8
+UnsignedInteger <-- U16
+UnsignedInteger <-- U32
+UnsignedInteger <-- U64
+UnsignedInteger <-- U128
+UnsignedInteger <-- USize
+UnsignedInteger <-- ULong
+
+SignedInteger <-- I8
+SignedInteger <-- I16
+SignedInteger <-- I32
+SignedInteger <-- I64
+SignedInteger <-- I128
+SignedInteger <-- ISize
+SignedInteger <-- ILong
+```
+
+## Methods of Concern
+
+```pony
+trait val Real[A: Real[A] val] is
+  (Stringable & _ArithmeticConvertible & Comparable[A])
+  ...
+  new val min_value()
+  new val max_value()
+  ...
+```
+
+`Real` will exist above `Rational`, `BigInt`, and `BigFloat` and as such would require defining the above methods for these types. `Rational` can be defined as minimum and maximum of the numerator, however `BigInt` and `BigFloat` by definition have arbitrary precision making defining a minimum and maximum difficult at the least -- if we define them as the minimum and maximum of a machine-sized int and float, or define them as -Inf and Inf -- or impossible at the worst -- if we define them by their possible limits which are arbitrary.
+
+```pony
+trait val Integer[A: Integer[A] val] is Real[A]
+  ...
+  fun op_and(y: A): A => this and y
+  fun op_or(y: A): A => this or y
+  fun op_xor(y: A): A => this xor y
+  fun op_not(): A => not this
+
+  fun bit_reverse(): A
+    """
+    Reverse the order of the bits within the integer.
+    For example, 0b11101101 (237) would return 0b10110111 (183).
+    """
+
+  fun bswap(): A
+```
+
+`Integer` will exist above `BigInt` and as such would require defining the above methods -- however `BigInt` will be defined via other numerics so these methods could be applied recursively.
```

Comment:
I don't understand what this means. What are "the above methods"? Those listed on the Integer trait? 

"`BigInt` will be defined via other numerics so these methods could be applied recursively" <-- I don't understand what this means either. There seems to be something that I am supposed to understand about "other numerics" that BigInt will be defined via (did I miss that somewhere, if yes, bringing information forward so that you don't have to hold large chunks of RFC in your head would be good). I'm also not sure what "applied recursively" means in this context.